Implementing license header checking in CI

john-science opened the pull request:
## Purpose

- To implement license header checking in CI

## Proposed Changes

What does this PR add, remove or fix?

I have added one GitHub Action CI (to check that every Python file has a Copyright header at the top). And I had to add one metadata file to the root of the repo, to make this CI action run properly.

## Issues

- To close #182 

## Testing

- How was it tested, what were the results?

This is it's own CI, so it is self-testing. After running it the first time, I found three `__init__.py` files that were lacking the header. Kind of an oddity, but looking around at some other open-source projects that have this license header rule (e.g. [armi](https://github.com/terrapower/armi/), this appears to be the standard, so I went with it.

john-science committed:
Adding license checker to CI

john-science committed:
Filling in missing copyright headers

maciejmajek commented:
Hi, Thank you! I don't think this is mentioned in the readmes, we use pre-commit for code formatting. Please run the following:
```bash
pre-commit run -a
```

adamdbrw commented:
Thank you for this contribution!

john-science commented:
> Hi, Thank you! I don't think this is mentioned in the readmes, we use pre-commit for code formatting. Please run the following:
> 
> ```shell
> pre-commit run -a
> ```

Oh, sorry about that.

Okay, I ran `precommit run -a` and it came back clean.  (After I fixed the linting errors.)

adamdbrw commented on .licenserc.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "**/*.py": "# Copyright "
```
```suggestion
  "**/*.py": [
  "# Copyright ",
  "#",
  "# Licensed under the Apache License, Version 2.0 (the \"License\");"
  ],
  "**/*.{h,cpp,hpp}": [
  "// Copyright ", 
  "//",
  "// Licensed under the Apache License, Version 2.0 (the \"License\");"
  ]
```

adamdbrw replied:
Rationale: 
We have C++ as well as Python in the project
We should not allow to include non-Apache Licenses in our core (we can add 3rd party src folder and exclude it from this requirement to allow for compatible licenses such as MIT, BSD 3, Mozilla later).

maciejmajek replied:
I've tested it using `npx @kt3k/license-checker`
with config
```json
{
    "**/*.py": [
        "# Copyright ",
        "#",
        "# Licensed under the Apache License, Version 2.0 (the \"License\");"
    ],
    "**/*.{h,cpp,hpp}": [
        "// Copyright ",
        "//",
        "// Licensed under the Apache License, Version 2.0 (the \"License\");"
    ]
}
```
and it seems to be working as expected

adamdbrw replied:
Great! How about you either apply + modify my suggestion or submit your patch (tested) with my co-authorship? I guess there are whitespace differences.

maciejmajek replied:
Up to @john-science

john-science replied:
Yeah, that change is solid. Taken.

I was only looking at the Python code, my bad.

adamdbrw reviewed (commented):
Checker suggestions

john-science committed:
Taking C++ improvements from adamdbrw

Thanks to @adamdbrw for the correction; I was only looking at the Python code.

Co-authored-by: Adam Dąbrowski <adam.dabrowski@robotec.ai>

maciejmajek reviewed (commented):
Please run the pre-commit 🙇🏼‍♂️, otherwise LGTM

john-science committed:
Fixing linting from suggestion

john-science commented:
> Please run the pre-commit 🙇🏼‍♂️, otherwise LGTM

Okay, linter came back clean.

maciejmajek reviewed (approved):
LGTM. Thanks!
